File: Week1/6_Majority_Element/Solution.py
#
# Created on Wed May 06 2020
#
# Title: Leetcode - Majority Element
#
# Author: Vatsal Mistry
# Web: mistryvatsal.github.io
#

# Problem solution using hashmap.
# Time complexity : O(n)
# Space complexity : O(n)
class Solution:
    def majorityElement(self, nums):
        store = dict()
        for i in nums:
            if i not in store:
                store[i] = 1
            else:
                store[i] += 1
        n_by_2 = len(nums) // 2
        result = 0
        for key, val in store.items():
            if val > n_by_2:
                result = key
        return result

File: Week1/6_Majority_Element/Solution.cpp
/*
 * Created on Wed May 06 2020
 *
 * Title: Leetcode - Majority Element
 *
 * Author: Vatsal Mistry
 * Web: mistryvatsal.github.io
 */

// Solution using hashmap datastructure, in cpp using unordered_map
// Time complexity : O(n)
// Space complexity : O(n)

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
    public:
    unordered_map<int, int> store;
    int majorityElement(vector<int>& nums) {
        for(auto i: nums) {
            store[i]++;
        }
        
        int n_by_2=nums.size() / 2, result=0;
        for(auto it=store.begin(); it!=store.end(); ++it) {
            if(it->second > n_by_2) {
                result = it->first;
            }
        }
        return result;
    }
};
